core/src/test/scala/org/apache/spark/ExecutorAllocationManagerSuite.scala

```diff
@@ -210,22 +216,282 @@ class ExecutorAllocationManagerSuite
+    val stage2 = createStageInfo(2, 20)
+    sc.listenerBus.postToAll(SparkListenerJobStart(1, 0, Seq(stage2), sc.getLocalProperties))
+    sc.listenerBus.postToAll(SparkListenerStageSubmitted(stage2))
+    assert(maxNumExecutorsNeeded(manager) === 16) // 5 + 11
+
+    // Submit a job in group 3
+    sc.setJobGroup("group3", "", false)
+    val stage3 = createStageInfo(3, 50)
+    sc.listenerBus.postToAll(SparkListenerJobStart(2, 0, Seq(stage3), sc.getLocalProperties))
+    sc.listenerBus.postToAll(SparkListenerStageSubmitted(stage3))
+    assert(maxNumExecutorsNeeded(manager) === 33) // 5 + 11 + 17
+
+    // Mark job in group 2 as complete
+    sc.listenerBus.postToAll(SparkListenerStageCompleted(stage2))
+    sc.listenerBus.postToAll(SparkListenerJobEnd(1, 20, JobSucceeded))
+    assert(maxNumExecutorsNeeded(manager) === 22) // 33 - 11
+
+    // Mark job in group 1 as complete
+    sc.listenerBus.postToAll(SparkListenerStageCompleted(stages(0)))
+    sc.listenerBus.postToAll(SparkListenerStageCompleted(stages(1)))
+    sc.listenerBus.postToAll(SparkListenerJobEnd(0, 10, JobSucceeded))
+    assert(maxNumExecutorsNeeded(manager) === 17) // 22 - 5
+
+    // Submit a job without any job group
+    sc.clearJobGroup()
+    val stage4 = createStageInfo(4, 333)
+    sc.listenerBus.postToAll(SparkListenerJobStart(4, 0, Seq(stage4), sc.getLocalProperties))
+    sc.listenerBus.postToAll(SparkListenerStageSubmitted(stage4))
+    assert(maxNumExecutorsNeeded(manager) === 350) // 17 + 333
+
+    // Mark job without job group as complete
+    sc.listenerBus.postToAll(SparkListenerStageCompleted(stage4))
+    sc.listenerBus.postToAll(SparkListenerJobEnd(4, 20, JobSucceeded))
+    assert(maxNumExecutorsNeeded(manager) === 17) // 350 - 333
+  }
+
+  test("add executors capped by max concurrent tasks for concurrent job groups with speculation") {
+    val conf = new SparkConf()
+      .setMaster("myDummyLocalExternalClusterManager")
+      .setAppName("test-executor-allocation-manager")
+      .set("spark.dynamicAllocation.enabled", "true")
+      .set("spark.dynamicAllocation.testing", "true")
+      .set("spark.job.group1.maxConcurrentTasks", "5")
+      .set("spark.job.group2.maxConcurrentTasks", "11")
+      .set("spark.job.group3.maxConcurrentTasks", "17")
+    val sc = new SparkContext(conf)
+    contexts += sc
+
+    val manager = sc.executorAllocationManager.get
+
+    // Submit a job in group1
+    sc.setJobGroup("group1", "", false)
+    val stages = Seq(createStageInfo(0, 2), createStageInfo(1, 10))
+    // Submit the job and stage start/submit events
+    sc.listenerBus.postToAll(SparkListenerJobStart(0, 0, stages, sc.getLocalProperties))
+    sc.listenerBus.postToAll(SparkListenerStageSubmitted(stages(0)))
+
+    // Verify that we're capped at number of max concurrent tasks in the job group
+    assert(maxNumExecutorsNeeded(manager) === 2)
+
+    // submit a speculative task
+    sc.listenerBus.postToAll(SparkListenerSpeculativeTaskSubmitted(0))
```

Comment:
this works because its pending, send the task started event for this and I think it will show the wrong value due to getIncompleteTasksForStage not handling speculative.

I think it would be good to have more tests due the task started to make sure the pending vs running logic works properly.

Reply:
okay. I will add more unit tests to cover the scenario.